src/openassetio-core-c/include/openassetio/c/SimpleMap.h

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+
+#include <openassetio/c/export.h>
+
+#include "./StringView.h"
+#include "./namespace.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```

Comment:
[question] @meshula - we're curious to get your opinion on whether we should try to support pure C linking (i.e. headers) or whether having a C-compatible interface in the binary is sufficient?

We're raising this specifically now since this file (and others) has a contradiction, in that we're apparently supporting pure C via `#ifdef __cplusplus`, but then we're also `#include`ing C++-only headers (e.g. `cstddef`). 

We need to decide whether to continue supporting pure C headers or not.  

Off the top of my head the only reasons to have C-compatible headers are
* C projects linking against our library. Is that really likely?
* FFIs/bindings generators that are able to automate some boilerplate by parsing C headers (but such systems may well be able to parse C++ headers too).

Reply:
The pattern I use is

```cpp
#ifdef __cplusplus
// individually adorn functions for complete control
// rather than block externing everything
#define MYAPI_EXTERNC extern "C"
#else
#define MYAPI_EXTERNC
#endif
#ifdef _MSC_VER
... extra junk for dll export/import and gcc visibility
#define MYAPI_EXPORT
#endif
// include the c headers not the c++ headers, the fussiness around
// std:: in front of everything when it comes from cmath seems like
// unhelpful complication and compiler to compiler variation in defaults

#include <math.h>

MYAPI_EXPORT MYAPI_EXTERN
int my_func();
```

as far as binding generators go, you get C for free pretty much across the board in languages that natively parse C headers, and autobind, e.g. Rust, Zig, Swift, Objective C, all do this. C++ always implies extra work. Rust, Zig, Swift, Objective C, will all require manual C binding of the C++ interfaces, and the results tend to be highly unidiomatic to the target language.

Languages with well developed c++ generators usually also require one to go back in and create a "porcelain" layer so that the generated goobledygook is consumable by humans. This tends not to be an issue with C autogens, and it all stems from the fact that C++ has extra semantics around ctor/dtor and memory management and the resultant bridging tends to be opinionated and clumsy IMO.

Reply:
I like @meshula 's pattern. Macroising the symbol visibility is common and familiar and if there's going to a by a dynamic library with this stuff in, something similar will be needed! Adding the extern-"C"-ness optionally when using the header as C++ is a nice addition.

We should definitely be using the C headers for things in all the C-compatible headers:

```
#include <math.h>
```

can be compiled as C or C++ (as long as the compilers math.h works as it should do! I had to read the C++ spec to check... :smile: ).

```
#include <cmath>
```

This just automatically requires all users of the header to be C++ compilers... which kind of defeats the point of making the headers valid C. We could instead use C++ and then control the ABI (if necessary) with compiler specific attributes in the function prefix macros.

Reply:
Really great feedback, thanks!  I've created #337 to fix up the headers (since it's not just the PR where the mistake has been made :roll_eyes: ) and #338 to ensure we have a consistent and tested strategy for our C header structure.